docetl/dataset.py

```diff
@@ -0,0 +1,326 @@
+
+        return self._apply_parsing_tools(data)
+
+    def _apply_parsing_tools(self, data: List[Dict]) -> List[Dict]:
+        """
+        Apply parsing tools to the data.
+
+        Args:
+            data (List[Dict]): The data to apply parsing tools to.
+
+        Returns:
+            List[Dict]: The data with parsing tools applied.
+
+        Raises:
+            ValueError: If a parsing tool is not found or if an input key is missing from an item.
+        """
+        for tool in self.parsing:
+            input_key = tool["input_key"]
+            if tool["function"] in PARSING_TOOLS:
+                func = PARSING_TOOLS[tool["function"]]
+            elif (
+                self.user_defined_parsing_tool_map
+                and tool["function"] in self.user_defined_parsing_tool_map
+            ):
+                func = eval(
+                    self.user_defined_parsing_tool_map[tool["function"]].function_code
+                )
+            else:
+                raise ValueError(
+                    f"Parsing tool {tool['function']} not found. Please define it or use one of our existing parsing tools: {PARSING_TOOLS.keys()}"
+                )
+
+            output_key = tool["output_key"]
+            function_kwargs = tool.get("function_kwargs", {})
+            new_data = []
+
+            def process_item(item):
+                if input_key not in item:
+                    raise ValueError(f"Input key {input_key} not found in item: {item}")
+                result = func(item[input_key], **function_kwargs)
+                if isinstance(result, list):
+                    return [item.copy() | {output_key: res} for res in result]
+                else:
+                    return [item | {output_key: result}]
+
+            with ThreadPoolExecutor() as executor:
+                futures = [executor.submit(process_item, item) for item in data]
+                new_data = []
+                for future in as_completed(futures):
+                    new_data.extend(future.result())
+
+            data = new_data
+
+        return data
+
+    def sample(self, n: int, random: bool = True) -> List[Dict]:
+        """
+        Sample n items from the dataset.
+
+        Args:
+            n (int): Number of items to sample.
+            random (bool): If True, sample randomly. If False, take the first n items.
+
+        Returns:
+            List[Dict]: A list of n sampled items.
+
+        Raises:
+            ValueError: If the sample size is larger than the dataset size or if the file extension is unsupported.
+        """
+        if self.type == "memory":
+            import random as rd
+
+            data = self.path_or_data
+            if n > len(data):
+                raise ValueError(
+                    f"Sample size {n} is larger than dataset size {len(data)}"
+                )
+            sampled_data = rd.sample(data, n) if random else data[:n]
+            return self._apply_parsing_tools(sampled_data)
```

Comment:
Assuming that there was already some cached output of `self._apply_parsing_tools` (called with all items of the dataset), would sampling on that rather than the document items be a problem? I imagine that this could save some compute (and may be IO) parsing work.

Now, of course, this would mean that there would need to be some caching done for the `self._apply_parsing_tools`

Reply:
Good point. I think this can be done in another PR, i'll make an issue for it